book/src/pong-tutorial/pong-tutorial-04.md

```diff
@@ -0,0 +1,351 @@
+                    } else if paddle.side == Side::Right && ball.velocity[0] > 0.0 {
+                        ball.velocity[0] = -ball.velocity[0];
+                    }
+                }
+            }
+        }
+    }
+}
+
+// A point is in a box when its coordinates are smaller or equal than the top
+// right, but larger or equal than the bottom left.
+fn point_in_rect(x: f32, y: f32, left: f32, bottom: f32, right: f32, top: f32) -> bool {
+    x >= left && x <= right && y >= bottom && y <= top
+}
+```
+
+This system is responsible for detecting collisions between balls and
+paddles, as well as balls and the top and bottom edges of the arena.
+If a collision is detected, the ball bounces off. This is done
+by modyfing the velocity of the `Ball` component.
+
+Also, don't forget to add `mod move_balls` and `mod bounce` in `systems/mod.rs`
+as well as adding our new systems to the game data:
+```rust,no_run,noplaypen
+# extern crate amethyst;
+# use amethyst::prelude::*;
+# use amethyst::core::transform::TransformBundle;
+# use amethyst::renderer::{DisplayConfig, DrawFlat, Pipeline,
+#                        PosTex, RenderBundle, Stage};
+# fn main() -> amethyst::Result<()> {
+# let path = "./resources/display_config.ron";
+# let config = DisplayConfig::load(&path);
+# let pipe = Pipeline::build().with_stage(Stage::with_backbuffer()
+#       .clear_target([0.0, 0.0, 0.0, 1.0], 1.0)
+#       .with_pass(DrawFlat::<PosTex>::new()),
+# );
+# mod systems {
+# use amethyst;
+# pub struct PaddleSystem;
+# impl<'a> amethyst::ecs::System<'a> for PaddleSystem {
+# type SystemData = ();
+# fn run(&mut self, _: Self::SystemData) { }
+# }
+# pub struct MoveBallsSystem;
+# impl<'a> amethyst::ecs::System<'a> for MoveBallsSystem {
+# type SystemData = ();
+# fn run(&mut self, _: Self::SystemData) { }
+# }
+# pub struct BounceSystem;
+# impl<'a> amethyst::ecs::System<'a> for BounceSystem {
+# type SystemData = ();
+# fn run(&mut self, _: Self::SystemData) { }
+# }
+# }
+# let input_bundle = amethyst::input::InputBundle::<String, String>::new();
+let game_data = GameDataBuilder::default()
+#    .with_bundle(RenderBundle::new(pipe, Some(config)).with_sprite_sheet_processor())?
+#    .with_bundle(TransformBundle::new())?
+#    .with_bundle(input_bundle)?
+#    .with(systems::PaddleSystem, "paddle_system", &["input_system"])
+    // ...other systems...
+    .with(systems::MoveBallsSystem, "ball_system", &[])
+    .with(
+        systems::BounceSystem,
+        "collision_system",
+        &["paddle_system", "ball_system"],
+    );
+# Ok(())
+# }
+```
+
+You should now have a ball moving and bouncing off paddles and off the top and bottom
+of the screen. However, you will quickly notice that if the ball goes out of the screen
+on the right or the left, it never comes back and the game is over...
+
+## Summary
+
+In this chapter, we finally added a ball to our game.
+In the next chapter, we'll add a system that checks when ball
```

Comment:
"we'll add a system checking when a player loses the game, and add a scoring system!"